src/Storages/MergeTree/DataPartStorageOnDisk.cpp

```diff
@@ -322,12 +324,19 @@ void DataPartStorageOnDisk::clearDirectory(
     const NameSet & names_not_to_remove,
     const MergeTreeDataPartChecksums & checksums,
     const std::unordered_set<String> & skip_directories,
+    bool is_temp,
+    MergeTreeDataPartState state,
     Poco::Logger * log,
     bool is_projection) const
 {
     auto disk = volume->getDisk();
 
-    if (checksums.empty())
+    /// It does not make sense to try fast path for incomplete temporary parts, because some files are probably absent.
+    /// Sometimes we add something to checksums.files before actually writing checksums and columns on disk.
+    /// Also sometimes we write checksums.txt and columns.txt in arbitrary order, so this check becomes complex...
+    bool looks_like_temporary_part = is_temp || state == MergeTreeDataPartState::Temporary;
```

Comment:
why "looks like"?)) Can we have `is_temp` or `state == Temporary` for non-temporary part?

Reply:
The problem is that I forgot again why do we have both `is_temp` flag and `Temporary` state...